greenfoot/src/main/java/greenfoot/sound/SoundPreferencePanel.java

```diff
@@ -0,0 +1,166 @@
+{
+    private final ComboBox<Mixer> inputDevice = new ComboBox<>();
+    private final ComboBox<Mixer> outputDevice = new ComboBox<>();
+    
+    public SoundPreferencePanel()
+    {
+        // Initialise available items first:
+        
+        // This is the right way round; target lines are input devices, source lines are output devices:
+        initialize(inputDevice, true, mixer -> Arrays.stream(mixer.getTargetLineInfo()).anyMatch(i -> i.getLineClass().equals(TargetDataLine.class)));
+        initialize(outputDevice, false, mixer -> Arrays.stream(mixer.getSourceLineInfo()).anyMatch(i -> i.getLineClass().equals(SourceDataLine.class)));
+        
+        Label inputLabel = new Label(Config.getString("greenfoot.sound.input"));
+        add(inputLabel, 0, 0);
+        add(inputDevice, 1, 0);
+        Label outputLabel = new Label(Config.getString("greenfoot.sound.output"));
+        add(outputLabel, 0, 1);
+        add(outputDevice, 1, 1);
+        Label warning = new Label(Config.getString("greenfoot.sound.warning"));
+        warning.setWrapText(true);
+        add(warning, 0, 2);
+        setColumnSpan(warning, 2);
+        setVgap(7);
+        setHgap(5);
+        ColumnConstraints cc0 = new ColumnConstraints();
+        cc0.setMinWidth(GridPane.USE_PREF_SIZE);
+        cc0.setHgrow(Priority.ALWAYS);
+        ColumnConstraints cc1 = new ColumnConstraints();
+        cc1.setMinWidth(GridPane.USE_PREF_SIZE);
+        getColumnConstraints().addAll(cc0, cc1);
+    }
+
+    /**
+     * Initialize the given combo box with the list of available mixers
+     * 
+     * @param audioDropdown The combo box to add the items to.  The selection will be loaded from the preferences file, or system default if blank
+     * @param input Whether this an input device (true) or output device (false)
+     * @param valid The test for whether the mixer is a valid option to add.  Some mixers are things like volume controls, not actual input or output devices.
+     */
+    private static void initialize(ComboBox<Mixer> audioDropdown, boolean input, Predicate<Mixer> valid)
+    {
+        audioDropdown.getItems().clear();
+        // We always add null, which is the system default:
+        audioDropdown.getItems().add(null);
+        for (Mixer.Info mixerInfo : AudioSystem.getMixerInfo())
+        {
+            Mixer mixer = AudioSystem.getMixer(mixerInfo);
+            if (valid.test(mixer))
+            {
+                audioDropdown.getItems().add(mixer);
+            }
+        }
+        
+        // Load item from preferences file:
+        audioDropdown.getSelectionModel().select(SoundUtils.loadMixer(audioDropdown.getItems(), input));
+
+        audioDropdown.setConverter(new StringConverter<Mixer>()
+        {
+            @Override
+            public String toString(Mixer object)
+            {
+                return object == null ? "Default" : object.getMixerInfo().getName();
```

Comment:
I wonder if "Default" can mislead users to think it would be the one they chose once they change mixer (as "default to what Greenfoot is using now"). Do you think it is better using "System default"?

Reply:
Perhaps "Java default"?  Confusingly, one of the options available on Windows is something like "Primary sound driver" although this doesn't necessarily seem to be the default choice by Java, even though it is something like the System default.